CARDS-1962/1963: Add configurable PREMs import exclusion and clinic mapper

acrowthe opened the pull request:
Adds a configurable exclusion and clinic mapper to the clarity import process.

Configuration is via OSGI config.

Testing can be performed by following testing instructions in [CARDS-1780](https://github.com/data-team-uhn/cards/pull/1204) using `processorImportTestSql.sql` instead of `importTestSql.sql`

By default, patients should have their visits assigned to the following cohorts:
- 1: IP
- 7: Rehab
- 8, 9: ED
- 14: EDIP
- 15: IP
- 18: IP
- 20: ED
- 21: EDIP
- 27: IP
- 31: Rehab


The other patients should have their visits discarded. This can be changed by adjusting the column values in `processorImportTestSql` to test different conditions

[CARDS-1780]: https://phenotips.atlassian.net/browse/CARDS-1780?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

acrowthe committed:
CARDS-1963/1963: Add configurable PREMs import exclusion and clinic mapper

acrowthe committed:
CARDS-1963/1963: Add configurable PREMs import exclusion and clinic mapper
Add testing sql file

sdumitriu commented on prems-resources/backend/src/main/java/io/uhndata/cards/prems/internal/importer/ConfiguredCohortMapper.java:
```diff
@@ -0,0 +1,81 @@
+import io.uhndata.cards.clarity.importer.spi.ClarityDataProcessor;
+
+/**
+ * Clarity import processor that assigns a cohort to an imported visit based on OSGI configured conditons.
```
```suggestion
 * Clarity import processor that assigns a cohort to an imported visit based on OSGi configured conditions.
```

sdumitriu commented on prems-resources/backend/src/main/java/io/uhndata/cards/prems/internal/importer/AbstractConditionalClarityDataProcessor.java:
```diff
@@ -0,0 +1,183 @@
+    private class ConditionDefinition
+    {
+        private final String[] dualOperators = {"<>", "<=", "<", ">=", ">", "="};
+        private final String[] singleOperators = {"is empty", "is not empty"};
```
I think the standard terms are `binaryOperators` and `unaryOperators`. Also, they could be static, so `BINARY_OPERATORS` and `UNARY_OPERATORS`.

But combined with the cyclomatic complexity (e.g. too much code) in the `matches` method, this could be turned into an enum, each member doing its own bit of checking.

sdumitriu replied:
Done that in fa7a00d41

sdumitriu commented on prems-resources/backend/src/main/java/io/uhndata/cards/prems/internal/importer/ConfiguredCohortMapper.java:
```diff
@@ -0,0 +1,81 @@
+ * @version $Id$
+ */
+@Designate(ocd = ConfiguredCohortMapper.Config.class, factory = true)
+@Component(configurationPolicy = ConfigurationPolicy.REQUIRE)
```
It is not needed to specify `configurationPolicy`, since by default it is set to `REQUIRE` if `factory=true`.

sdumitriu commented on prems-resources/backend/src/main/java/io/uhndata/cards/prems/internal/importer/ConfiguredCohortMapper.java:
```diff
@@ -0,0 +1,81 @@
+        + " cohort")
+    public static @interface Config
+    {
+        @AttributeDefinition(name = "Priority", description = "Priority")
```
Having the same description as the name doesn't add value, so since having a description is optional, either remove it, or set a more descriptive description.

sdumitriu commented on prems-resources/backend/src/main/java/io/uhndata/cards/prems/internal/importer/ConfiguredCohortMapper.java:
```diff
@@ -0,0 +1,81 @@
+        @AttributeDefinition(name = "ClinicId",
+            description = "Clinic mapping path that should be assigned if all conditions are met"
+                + " (eg. /Survey/ClinicMapping/123456789)")
+        String clinicId();
```
Since the expected value is a path to a clinic, not just its identifier, a better and simpler name is just `clinic`.

sdumitriu commented on prems-resources/backend/src/main/java/io/uhndata/cards/prems/internal/importer/AbstractConditionalClarityDataProcessor.java:
```diff
@@ -0,0 +1,183 @@
+                    break;
+                case "=":
+                    // Return true if column has entry and values match
+                    result = inputValue != null && inputValue.matches(this.value);
```
I think we may have an issue with checking for equality with a number, e.g if the filter is on `SOME_COLUMN = 1.5`, it may also match `125`. We have to be careful when comparing such numbers to escape any eventual dots in the value.

Or perhaps add a specific `[not] matches` or `[not] like` operator, and leave `=` as exact match comparison?

acrowthe replied:
Added `not matches` and `matches`, changed `=` and `<>` to be basic string compares and matches to be regex based

sdumitriu commented on prems-resources/backend/src/main/java/io/uhndata/cards/prems/internal/importer/ConfiguredCohortMapper.java:
```diff
@@ -0,0 +1,81 @@
+        String clinicId();
+
+        @AttributeDefinition(name = "Conditions", description = "Conditions for this cohort to be assigned."
+            + " For example \"COLUMN_NAME is empty\".")
```
Should we list all supported operators?

sdumitriu commented on prems-resources/backend/src/main/java/io/uhndata/cards/prems/internal/importer/SendCPESForDepartmentFrequency.java:
```diff
@@ -74,7 +74,8 @@ public SendCPESForDepartmentFrequency(SendCPESForDepartmentFrequencyConfigDefini
     {
         final String department = input.get("DISCH_DEPT_NAME");
-        if (Math.random() < this.perDepartmentFrequency.getOrDefault(department, this.defaultFrequency)) {
+        if (input.getOrDefault("CLINIC", "").length() == 0
```
@marta-  Wasn't CPESIC supposed to randomly apply to any visit assigned to a survey?

sdumitriu replied:
Actually, not for ED, just IP.

sdumitriu commented on prems-resources/feature/src/main/features/feature.json:
```diff
@@ -179,6 +179,66 @@
+    "io.uhndata.cards.prems.internal.importer.ConfiguredDiscardFilter~UHN-CohortMapper-ED-TransferTW":{
+      "priority": 100,
+      "clinicId": "/Survey/ClinicMapping/-1792626799",
+      "conditions": ["CLINIC is empty", "DISCH_DEPT_NAME = TW-EMERGENCY"]
```
Maybe I'm not reading this properly, but doesn't `ED-TransferTW` make `ED-NoTransferTW` useless? Or will there be more mappers added in between these two that may change which survey is assigned?

acrowthe replied:
Fixed logic around ED

sdumitriu commented on prems-resources/feature/src/main/features/feature.json:
```diff
@@ -179,6 +179,66 @@
+      "priority": 100,
+      "conditions": ["CLINIC is empty", "LENGTH_OF_STAY_DAYS <= 4"]
+    },
+    "io.uhndata.cards.prems.internal.importer.ConfiguredDiscardFilter~UHN-FilteredDiscard-CPES-IP-Psychiatric":{
```
Why the `CPES` in the name in the above 2 rules?

sdumitriu commented on prems-resources/feature/src/main/features/feature.json:
```diff
@@ -179,6 +179,66 @@
+    },
+    "io.uhndata.cards.prems.internal.importer.ConfiguredDiscardFilter~UHN-FilteredDiscard-CPES-IP-Psychiatric":{
+      "priority": 100,
+      "conditions": ["CLINIC is empty", "DISCH_DEPT_NAME = TG-8ES"]
```
As I see in the specs, this should apply to all events, including ED, so the `clinic is empty` rule should be removed.

sdumitriu committed:
CARDS-1963: Add configurable PREMs import exclusion and clinic mapper
Refactoring: made the operator code more flexible

sashaandjic commented:
@acrowthe  I am not getting anything imported.
Here are the first steps that seemed to work properly:
![image](https://user-images.githubusercontent.com/23243244/218512143-380cf5e0-818f-4f33-a025-29e4e01836b0.png)
![image](https://user-images.githubusercontent.com/23243244/218512180-4f799b92-f4eb-4885-9f65-517160f03dc9.png)

acrowthe committed:
CARDS-1963-1963: Add test cases for merged filters/mappers

acrowthe committed:
CARDS-1963-1963: Add configurable clarity import mapper/filters
Address code review comments:
- Add seperate plaintext and regex comparisons and adjust configurations to match
- Improve OSGi config descriptions and names
- Adjust configuration priorities
- Correctly discard all transferred ED visits
- Fix potention NPE in CPES mapper if clinic is configured wrong

acrowthe committed:
CARDS-1963-1963: Add configurable clarity import mapper/filters
- Clean up unneeded configurationPolicy

sashaandjic commented:
I am having the same problem - no import
![image](https://user-images.githubusercontent.com/23243244/218712801-04604fbe-03d4-426c-b1b2-d84f0b4068ed.png)
![image](https://user-images.githubusercontent.com/23243244/218712837-4c55150a-2499-4896-9de2-987e25807d5a.png)

sashaandjic reviewed (approved):
tested : 9 patients & 9 visits imported
P1 - IP
P7 - Rehab
P8 - ED
P9 - ED
P14- ED+IP
P16 - IP
P18 - IP
P20 - ED
P21 - ED+IP
![image](https://user-images.githubusercontent.com/23243244/218715687-61e718bc-6213-4479-a092-6778126e9adb.png)
![image](https://user-images.githubusercontent.com/23243244/218715789-220c5041-b890-49c8-9bad-612027a47a24.png)
![image](https://user-images.githubusercontent.com/23243244/218715835-0b5f4c57-3b67-4b40-b9de-e5833f62b91c.png)

sashaandjic commented:
QUESTION: no SPESIC survey?

marta- commented:
@marta- please note the thick line below ED in the last screenshot. For some reason this thick line always appears after ED.

That's a horizontal scrollbar caused by the fact that the ED title is so long.

marta- commented:
> **QUESTION**: no SPESIC survey?

That one is assigned randomly to 4% of the discharges. With so few test cases, it's very unlikely to get it. For testing purposes, the percentage can be increased drastically (>50%) in the OSGi config (see instructions here: https://github.com/data-team-uhn/cards/pull/1288) and hopefully a couple of CPESICs will pop up then.

marta- commented on prems-resources/feature/src/main/features/feature.json:
```diff
@@ -179,6 +179,64 @@
+    },
+    "io.uhndata.cards.prems.internal.importer.ConfiguredDiscardFilter~UHN-Discard-IP-Psychiatric":{
+      "priority": 20,
+      "conditions": ["DISCH_DEPT_NAME = TG-8ES"]
```
`= TG-8ES PSYCHIATRY` or `matches TG-8ES .*`

acrowthe committed:
CARDS-1962: Add 'in' and 'not in' mapper/filter function

acrowthe committed:
CARDS-1962: Add additional configured discard filters and tests

sdumitriu commented on prems-resources/feature/src/main/features/feature.json:
```diff
@@ -186,21 +186,41 @@
     "io.uhndata.cards.prems.internal.importer.ConfiguredDiscardFilter~UHN-Discard-Deceased":{
       "priority": 20,
-      "conditions": ["DISCH_DISPOSITION = Deceased"]
+      "conditions": ["DISCH_DISPOSITION in Not Arrived; Jail; Cad Donor; Deceased; Died on Leav; Diverted; Died; Donor; Stillbirth; Suicide Out; Death After; MAID; Death on Arr; IP Medically; IP In-Facili; IP Died Whil; IP Out of Fa; DOA; OP Medically; OP In-Facili; Suicide fac; Still Born; Shelter; Cad donor; Pt. suicide; Expired LOA; Expired; Jail or half; Med assist d; Suicide fac; DOA; Inpati ...
```
Died on Leav or Died on Leave?

Huh, are these standard values? I guess there's a length limit of 12 chars.

marta- replied:
Yeah, they're "standard"... and indeed there's a length limit.

sdumitriu commented on prems-resources/feature/src/main/features/feature.json:
```diff
@@ -186,21 +186,41 @@
+    },
+    "io.uhndata.cards.prems.internal.importer.ConfiguredDiscardFilter~UHN-Discard-Rehab-ComplexContinuing":{
+      "priority": 60,
+      "conditions": ["CLINIC = /Survey/ClinicMapping/78840662", "DISCH_DEPT_NAME in BC-3 NORTH TCU IP; BC-5A NORTH IP; BC-5B NORTH IP; BC-3 SOUTH IP; BC-5 SOUTH IP"]
```
Technically this could be merged with the previous one, but I think it's better to keep them separate as two different business logic decisions.

marta- replied:
More exclusions here: anything containing 'CCC'

marta- commented on prems-resources/feature/src/main/features/feature.json:
```diff
@@ -179,6 +179,84 @@
+      "priority": 20,
+      "conditions": ["DEATH_DATE is not empty"]
+    },
+    "io.uhndata.cards.prems.internal.importer.ConfiguredDiscardFilter~UHN-Discard-Deceased":{
+      "priority": 20,
+      "conditions": ["DISCH_DISPOSITION in Not Arrived; Jail; Cad Donor; Deceased; Died on Leav; Diverted; Died; Donor; Stillbirth; Suicide Out; Death After; MAID; Death on Arr; IP Medically; IP In-Facili; IP Died Whil; IP Out of Fa; DOA; OP Medically; OP In-Facili; Suicide fac; Still Born; Shelter; Cad donor; Pt. suicide; Expired LOA; Expired; Jail or half; Med assist d; Suicide fac; DOA; Inpati ...
+    },
```
Not all the conditions match this filter's title (e.g. Diverted, Shelter, Jail, ...). For the sake of semantics, let's make another filter `Discard-DichargeToLocation`

marta- replied:
New values to be added in DischargeToLocation: IP Trnsfr, OP Transfer.

sdumitriu commented on prems-resources/backend/src/main/java/io/uhndata/cards/prems/internal/importer/AbstractConditionalClarityDataProcessor.java:
```diff
@@ -0,0 +1,182 @@
+                    return Pair.of(StringUtils.removeEnd(configuration, this.operator).trim(), null);
+                }
+            } else {
+                String[] pieces = configuration.split("\\s+" + this.operator + "\\s+", 2);
```
This now requires spaces around all operators. We could instead put spaces only around the operators that need them, e.g. `NEQ("<>"...` but `IN(" in "...`, and change this regexp to:

```suggestion
                String[] pieces = configuration.split("\\s*" + this.operator + "\\s*", 2);
```

marta- commented on prems-resources/feature/src/main/features/feature.json:
```diff
@@ -179,6 +179,84 @@
       "emailConfiguration": "/apps/cards/clinics/UHN-Rehab/mailTemplates/ReminderNotification",
       "daysToVisit": -11
+    },
+    "io.uhndata.cards.prems.internal.importer.ConfiguredDiscardFilter~UHN-Discard-DeathDate":{
```
I'm not sure why every filter or mapper name is prefixed with `UHN-`. If there's no good reason for that prefix, I'd say we drop it.

sdumitriu commented on prems-resources/backend/src/main/java/io/uhndata/cards/prems/internal/importer/AbstractConditionalClarityDataProcessor.java:
```diff
@@ -0,0 +1,182 @@
+        GTE(">=", false, (input, value) -> compareDouble(input, value, (a, b) -> a >= b)),
+        GT(">", false, (input, value) -> compareDouble(input, value, (a, b) -> a > b)),
+        EQ("=", false, (input, value) -> input != null && input.equalsIgnoreCase(value)),
+        NOT_MATCHES("not matches", false, (input, value) -> input == null || !input.matches(value)),
+        MATCHES("matches", false, (input, value) -> input != null && input.matches(value)),
+        NOT_IN("not in", false, (input, value) -> input == null
+            && !Arrays.stream(value.split("\\s*;\\s*")).anyMatch(input::equalsIgnoreCase)),
+        IN("in", false, (input, value) -> input != null
+            && Arrays.stream(value.split("\\s*;\\s*")).anyMatch(input::equalsIgnoreCase)),
+        EMPTY("is empty", true, (input, value) -> input == null || input.length() == 0),
+        NOT_EMPTY("is not empty", true, (input, value) -> input != null && input.length() > 0);
```
```suggestion
        NOT_MATCHES(" not matches ", false, (input, value) -> input == null || !input.matches(value)),
        MATCHES(" matches ", false, (input, value) -> input != null && input.matches(value)),
        NOT_IN(" not in ", false, (input, value) -> input == null
            && !Arrays.stream(value.split("\\s*+;\\s*+")).anyMatch(input::equalsIgnoreCase)),
        IN(" in ", false, (input, value) -> input != null
            && Arrays.stream(value.split("\\s*+;\\s*+")).anyMatch(input::equalsIgnoreCase)),
        EMPTY(" is empty", true, (input, value) -> input == null || input.length() == 0),
        NOT_EMPTY(" is not empty", true, (input, value) -> input != null && input.length() > 0);
```

`*+` is a [possessive matcher](https://www.regular-expressions.info/possessive.html) that improves performance by grabbing as many spaces as possible and not backtracking into them when failing to match the rest of the expression.

sdumitriu replied:
It would improve performance if the regexp is compiled.

acrowthe replied:
Regex not implemented, strings and possesive matcher implemented

sdumitriu commented on prems-resources/backend/src/main/java/io/uhndata/cards/prems/internal/importer/AbstractConditionalClarityDataProcessor.java:
```diff
@@ -0,0 +1,182 @@
+        @Override
+        public String toString()
+        {
+            return this.column + " " + this.operator + (this.value != null ? " " + this.value : "");
```
With spaces around the operator names this can be simplified as:

```suggestion
            return this.column + this.operator + (this.value != null ? this.value : "");
```

sashaandjic reviewed (approved):
tested
10 patients/visits/surveys imported
5 CPESIC _(modified the percentage to from 00.4 to 0.8)_
1 Rehab
3 ED
1 IP

-----

P1: CPESIC
P7: Rehab
P8: ED
P9: ED
P14: CPESIC
P15: CPESIC
P18: CPESIC
P20: ED
P21: CPESIC
P27: IP

-----
This time none of ED+IP were imported.

acrowthe committed:
CARDS-1962/1963: Configurable clarity import mappers and filters
Adjust some configurations and tweak regex

marta- commented on prems-resources/feature/src/main/features/feature.json:
```diff
@@ -221,6 +221,92 @@
+    },
+    "io.uhndata.cards.prems.internal.importer.ConfiguredDiscardFilter~Discard-NotRehab-PrimaryDiagnosis":{
+      "priority": 60,
+      "conditions": ["CLINIC is empty", "DX_NAME in Acute delirium; Adjustment disorder; Adjustment disorder with mixed anxiety and depressed mood; Adjustment disorders, unspecified; Agitation; Alcohol dependence in remission; Alcohol intoxication; Alcohol use; Alcohol use disorder; Alcohol withdrawal; Alcoholism in recovery; Alzheimer disease; Anxiety; Anxiety about health; Anxiety and depressio ...
```
Add:
* Acute alcoholic intoxication
* Anoxic brain injury
* Drug-induced psychotic disorder
* Essential tremor
* Overdose
* Overdose of drug/medicinal substance
* Overdose of tricyclic antidepressants
* Polysubstance dependence
* Seizure
* Seizure disorder
* Seizures
* Tremor
* Vapes nicotine containing substance

Please keep the `DX_NAME` list on line 268 in alphabetical order to make it easier to read and update later on.

sdumitriu commented on prems-resources/feature/src/main/features/feature.json:
```diff
@@ -221,6 +221,92 @@
+    },
+    "io.uhndata.cards.prems.internal.importer.ConfiguredDiscardFilter~Discard-Palliative":{
+      "priority": 20,
+      "conditions": ["DISCH_DEPT_NAME in PM-PALLIATIVE CARE ONCOLOGY CLINIC; TW-PALLIATIVE CARE CLINIC; TG-PALLIATIVE CARE; PM-16P PALLIATIVE CARE"]
```
Would a `matches .* PALLIATIVE.*` be useful?

marta- replied:
Yes, although I'm torn between a regexp for condition simplicity and a list that is easy for non-devs to review.

sdumitriu commented on prems-resources/feature/src/main/features/feature.json:
```diff
@@ -221,6 +221,92 @@
+    "io.uhndata.cards.prems.internal.importer.ConfiguredCohortMapper~CohortMapper-ED-IPTransferShortStay":{
+      "priority": 80,
+      "clinic": "/Survey/ClinicMapping/-1792626799",
+      "conditions": ["CLINIC is empty", "ED_IP_TRANSFER_YN = yes", "LENGTH_OF_STAY_DAYS <= 4", "DISCH_DEPT_NAME <> TW-EMERGENCY", "DISCH_DEPT_NAME <> TG-EMERGENCY"]
```
Would `DISCH_DEPT_NAME not matches .*EMERGENCY.*` be a more flexible rule?

sdumitriu commented on prems-resources/feature/src/main/features/feature.json:
```diff
@@ -221,6 +221,92 @@
+    },
+    "io.uhndata.cards.prems.internal.importer.ConfiguredDiscardFilter~Discard-ED-TransferTW":{
+      "priority": 100,
+      "conditions": ["CLINIC is empty", "DISCH_DEPT_NAME = TW-EMERGENCY"]
```
Would `DISCH_DEPT_NAME matches .*EMERGENCY.*` work?

marta- replied:
You just made me realize we need a location filter so we don't include non-participating hospitals. Please add an early filter saying location must be one of: 
* Toronto General Hospital
* Toronto Western Hospital
* Princess Margaret Cancer Centre
* Toronto Rehab

After that yes, we can use regexp for emergency department detection.

sdumitriu committed:
[codestyle] Remove useless 'static' modifier

sdumitriu committed:
CARDS-1962/1963: Configurable clarity import mappers and filters
New filter: discard all events from non-participating hospitals

sdumitriu committed:
CARDS-1962/1963: Configurable clarity import mappers and filters
Add some comments to the big feature file

sdumitriu committed:
CARDS-1962/1963: Configurable clarity import mappers and filters
More DX_NAME filters

sdumitriu committed:
CARDS-1962/1963: Configurable clarity import mappers and filters
More general matching of emergency departments

marta- commented on prems-resources/feature/src/main/features/feature.json:
```diff
@@ -221,6 +231,118 @@
       "export.format": "csv"
+    },
+
+    // Clarity import filters and fillers
```
```suggestion
    // Clarity import filters and mappers
```

marta- commented on prems-resources/feature/src/main/features/feature.json:
```diff
@@ -221,6 +231,118 @@
+        "DISCH_LOC_NAME <> Toronto Rehab"
+      ]
+    },
+    // Discard dead patients
```
```suggestion
    // Discard deceased patients
```

marta- commented on prems-resources/feature/src/main/features/feature.json:
```diff
@@ -221,6 +231,118 @@
+      "priority": 20,
+      "conditions": ["DISCH_DISPOSITION in Cad Donor; Deceased; Died on Leav; Died; Donor; Stillbirth; Suicide Out; Death After; MAID; Death on Arr; IP Medically; IP In-Facili; IP Died Whil; IP Out of Fa; DOA; OP Medically; OP In-Facili; Suicide fac; Still Born; Pt. suicide; Expired LOA; Expired; Med assist d; Suicide fac; DOA"]
+    },
+    // Discard patients discharged to another institution (or not arrived at all)
```
```suggestion
    // Discard patients discharged to another institution or not arrived at all
```

sdumitriu committed:
CARDS-1962/1963: Configurable clarity import mappers and filters
Minor comment improvements

acrowthe committed:
CARDS-1962/1963: Improve rehab handling and comments

sashaandjic commented:
QUESTION: Shouldn't patient 20 be in ED-IP clinic?
![image](https://user-images.githubusercontent.com/23243244/219459526-696a81bb-f235-4efe-963a-e32a47c12a77.png)
![image](https://user-images.githubusercontent.com/23243244/219459604-295d85f0-5e36-447d-b696-943b75f7bce3.png)

acrowthe commented:
@sashaandjic 
> QUESTION: Shouldn't patient 20 be in ED-IP clinic?

20 gets mapped to ED since their visit is too short for the full IP survey - Admitted 20th, discharged 21st

sdumitriu committed:
CARDS-1962/1963: Configurable clarity import mappers and filters
Also include the configuration for CPESIC

sashaandjic reviewed (commented):
tested
